### ISSUE TITLE
Avoid return 503 after closing fastify

### ISSUE THREAD

**Issue by allevo** [fastify/fastify#1714]
## 🚀 Feature Proposal

Currently while fastify is shutting down, all APIs returns 503.
I would like to make this behaviour configurable. 

## Motivation

In Kubernetes, when a deployment is updated, the process receives a SIGTERM. After that, your container continues to receive incoming requests. With the current behaviour all those responses are 503, even if the process is able to serve them.

So, I cannot directly attach the SIGTERM event to `fastify.close` method.

## Example

```js
const fastify = require('fastify')({ returns503OnClose: false }) // default true
```

The changes involve those lines:
https://github.com/fastify/fastify/blob/901911c553bea62a21799ec7df34d5e36d2bacd0/fastify.js#L240-L250

where the  `router.closeRoutes()` is done only if the flag `returns503OnClose` is set to  `true`.

NB: According to [node documentation](https://nodejs.org/docs/latest/api/net.html#net_server_close_callback) the callback will be called only when all connections are closed, so no requests are discarded.

## Edit:
old conversation https://github.com/fastify/fastify/pull/886

**Comment by RafaelGSS** [comment]
Sounds good. But, maybe `default: false`? Because this problem sounds like a bug. Why this problem did not occur on dev `nodemon`?

**Comment by mcollina** [comment]
I think this sounds like a bug to me. Maybe we can fix it without a flag. Would the following work?

```js
 avvio.once('preReady', () => { 
   fastify.onClose((instance, done) => { 
     fastify[kState].closing = true 
     if (fastify[kState].listening) { 
       instance.server.close(function (err) {
          router.closeRoutes() 
          done(err)
       }) 
     } else { 
       router.closeRoutes() 

       done(null) 
     } 
   }) 
 }) 
```

**Comment by delvedor** [comment]
I'm sorry, but I'm missing why this functionality does not work for you.
The idea is that if you call `fastify.close`, Fastify should stop accepting new requests and try to close currently opened one.

Following the implementation:
https://github.com/fastify/fastify/blob/901911c553bea62a21799ec7df34d5e36d2bacd0/lib/route.js#L266-L282

**Comment by mcollina** [comment]
@delvedor what is happening is that we call `.close()`, however some calls are still getting in while the async close operation is happening. By moving the `router.closeRoutes()` and setting `closing` inside the callback, those request will actually get through. I've updated the snippet:

```js
avvio.once('preReady', () => { 
   fastify.onClose((instance, done) => { 
     if (fastify[kState].listening) { 
       instance.server.close(function (err) {
          router.closeRoutes()
          fastify[kState].closing = true  
          done(err)
       }) 
     } else { 
       router.closeRoutes() 
       fastify[kState].closing = true  
       done(null) 
     } 
   }) 
 })
```

**Comment by allevo** [comment]
The close method avoids new incoming connection.
If your application lives behind a proxy, your connection may be with 1.1 http version. This imply that all tcp connections are 'keep alive'. So the current connections can be used for new http requests. Currently all those requests receive 503

The next step, in my mind, is to make a plug-in that register a sigterm handler that invokes stop and forces to set the connection header to close

**Comment by mcollina** [comment]
> If your application lives behind a proxy, your connection may be with 1.1 http version. This implies that all tcp connections are 'keep alive'. So the current connections can be used for new http requests. Currently all those requests receive 503

That's the idea. 503 a usually handled and they mean "send this request somewhere else". Note that the shutdown sequence of fastify means that _we cannot guarantee_ that those requests could be handled at all, because a database connection could have been stopped for example.

Tracking all connections is also very expensive, and the framework has been designed in this way to have a safe/sane behavior without requiring that. If the current behavior has a problem, then we might want to change/improve that, as this seems to be a common use case.

**Comment by allevo** [comment]
Are you sure? 
Using the change mentioned before, I thought that all TCP connections are closed before any 'onClose' hook handlers are fired because the hook handlers are called in sequence. Or Im wrong?

**Comment by mcollina** [comment]
Calling .close on a server closes it, meaning there will be no new incoming TCP connections. It does nothing on all connected clients unfortunately.

**Comment by mcollina** [comment]
Actually I have been wrong. From https://nodejs.org/api/net.html#net_server_close_callback:

> Stops the server from accepting new connections and keeps existing connections. This function is asynchronous, the server is finally closed when all connections are ended and the server emits a 'close' event. The optional callback will be called once the 'close' event occurs. Unlike that event, it will be called with an Error as its only argument if the server was not open when it was closed.

Essentially when the `close() ` callback fires, there are no more connections lingering around.

@allevo would you mind sending a PR to disable the 503 behavior? We should still be sending the "connection: close" headers anyway on response

**Comment by allevo** [comment]
I'll fire asap

**Comment by allevo** [comment]
Fired!
https://github.com/fastify/fastify/pull/1723

**Comment by dgonzalez** [comment]
My 2 cents: on SIGTERM stop taking requests by returning the appropriate http code on the readiness probe (if you return anything >=300 k8s will stop sending you traffic). Then, wait for all the current requests to be processed and then, close the server. Would that work?

**Comment by delvedor** [comment]
Closed in #1723.
